framework/src/media/media_recorder.c

```diff
@@ -0,0 +1,343 @@
+
+record_result_t media_stop_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state == RECORD_PAUSED) {
+		g_rc.state = RECORD_STOPPING;
+		sem_post(&g_rc.sem_pause);
+	} else if (g_rc.state != RECORD_RUNNING) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_STOPPING;
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_pause_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_RUNNING) {
+		RECORD_UNLOCK();		
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_PAUSED;
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_resume_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_PAUSED) {
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_RUNNING;
+	media_record_prepare();	
+	sem_post(&g_rc.sem_pause);
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_record_init(void)
+{
+	if (!b_started) {
+		b_started = true;
+		memset(&g_rc, 0, sizeof(g_rc));
+
+		pthread_mutexattr_init(&g_rc.mutex_attr_record);
+		pthread_mutexattr_settype(&g_rc.mutex_attr_record, PTHREAD_MUTEX_RECURSIVE);
+		pthread_mutex_init(&g_rc.mutex_record, &g_rc.mutex_attr_record);
+
+		g_rc.fd = -1;
+		g_rc.state = RECORD_AVAILABLE;
+		g_rc.pcmin = NULL;
+		g_rc.pcmconf = NULL;
+
+		sem_init(&g_rc.sem_pause, 0, 0);
+	}
+	return RECORD_OK;
+}
+
+static bool write_full(void *ptr, size_t size)
+{
+	int ret;
+	size_t written = 0;
+	while (written < size) {
+		ret = write(g_rc.fd, (char *)ptr + written, size - written);
+		if (ret < 0) {
+			return false;
+		}
+		written += ret;
+	}
+	return true;
+}
+
+int record_and_write(char *buffer)
+{
+	int frames;
+	int size;
+	char inst = 0x05;
+
+	frames = pcm_readi(g_rc.pcmin, buffer, g_rc.buffer_size);
+
+	if (frames > 0) {
+		size = pcm_frames_to_bytes(g_rc.pcmin, frames);
+		if (IS_SOCKET(g_rc.fd))	{
+			write_full(&inst, 1);
+			write_full(&size, sizeof(int));
+		}
+		write_full(buffer, size);
+	}
+	return frames;
+}
+
+int threadfunc_media_recording(void *args)
+{
+	int ret;
+	char *buffer = NULL;
+	char inst = 0;
+	buffer = (char *)malloc(g_rc.buffer_size);
+
+	if (!buffer) {
+		RECORD_LOCK();
+		g_rc.state = RECORD_AVAILABLE;
+		RECORD_UNLOCK();
+		return RECORD_ERROR_MALLOC;
+	}
+
+	while (g_rc.thread_running) {
+		RECORD_LOCK();
+		switch (g_rc.state) {
+		case RECORD_RUNNING:
+			ret = record_and_write(buffer);
+			if (ret == 0) {
+				g_rc.state = RECORD_DROPPING;
+			} else if (ret < 0) {
+				g_rc.state = RECORD_STOPPING;
+			}
+			break;
+
+		case RECORD_PAUSED: {
+			RECORD_UNLOCK();
+			pcm_close(g_rc.pcmin);
+			g_rc.pcmin = NULL;
+			if (IS_SOCKET(g_rc.fd)) {
+				inst = 0x02;
+				write_full(&inst, 1);
+			}
+			sem_wait(&g_rc.sem_pause);
+			RECORD_LOCK();
+		}
+		break;
+
+		case RECORD_STOPPING: {
+			if (IS_SOCKET(g_rc.fd)) {
+				inst = 0x03;
+				write_full(&inst, 1);
+			}
+			close(g_rc.fd);
+			g_rc.fd = -1;
+			g_rc.thread_running = false;
+		}
+		break;
+
+		case RECORD_DROPPING:
```

Comment:
As we discuss many times, now time to decide it.
Actually all of pause and stop we will call pcm_drop.
But pause case, if we want start again, we should check manything.
1. record on file
  open same file again -> seek to last pos of file -> record & write again
2. send recorded data through network
  need more logic...